logger: allow starting before the driver is loaded

marc-hb opened the pull request:
See commit messages.

Fixes a small part of thesofproject/linux#3275

~Draft until it's been tested more.~ I used this for months

ujfalusi commented on tools/logger/logger.c:
```diff
@@ -78,7 +79,7 @@ static void usage(void)
 static int snapshot(const char *name)
 {
-	const char *path = "/sys/kernel/debug/sof";
+	const char *path = SYS_DEBUG "/sof";
```
it does not help readability, it was much cleaner when the whole path was humanly readable.

marc-hb replied:
Agreed but `sizeof("/sys/kernel/debug"/)` seemed ugly. I guess I can use a local variable there.

ujfalusi commented on tools/logger/logger.c:
```diff
@@ -362,6 +351,17 @@ int main(int argc, char *argv[])
+				config.version_file, strerror(ret));
+			goto out;
+		}
+	}
```
this is fine, but the commit message is somehow hard to digest:
"So when the driver is not loaded, we get the same (missing trace) error
trace message whether we use the -n option or not."

When the driver is not loaded then the entire `/sys/kernel/debug/sof` directory is non existent, so it does not matter which file you try to read, none of them will be present.

However, _if_ the dtrace is disabled in kernel then the `trace` and `filter` files are going to be missing and they will only appear if you reload the modules with `options snd_sof sof_debug=1` at minimum.

marc-hb replied:
> When the driver is not loaded then the entire /sys/kernel/debug/sof directory is non existent, so it does not matter which file you try to read, none of them will be present.

It matters a lot for the next commit but I tried to avoid a reference to the future.

> However, if the dtrace is disabled in kernel then the trace and filter files are going to be missing and they will only appear if you reload the modules with options snd_sof sof_debug=1 at minimum.

and? Not sure I see your point here sorry, is this a better commit message?

ujfalusi replied:
@marc-hb, what about printing something like "driver is not loaded" when the driver is not loaded (the sof debugfs directory is not present along with the fw_version file)?

If the version file is optional then why we fail when it is not present?

marc-hb replied:
> If the version file is optional then why we fail when it is not present?

We don't, this code is not run when config.version_fw is false.

I moved the version check _after_ waiting for the trace file. I left the error handling because it does not hurt, opening could fail for some other reason (like a race)

There is a "waiting for /sys/..." message above.

ujfalusi commented on tools/logger/logger.c:
```diff
@@ -171,6 +174,75 @@ static int append_filter_config(struct convert_config *config, const char *input
+	const int dwatch = inotify_add_watch(iqueue, watched_dir, IN_CREATE);
+	FILE *fstream;
+
+	char * const fpath = malloc(strlen(watched_dir) + 1 + strlen(expected_file) + 1);
```
can we have the declarations first?

marc-hb replied:
Why, do we need to support any legacy, pre-C99 compiler here? Leaving variables uninitialized makes the code more verbose and less safe: not possible to use `const` and more risk of errors. 
This is not the Linux kernel here, no need to support a baroque bad habit that never even existed in any other programming language.

https://stackoverflow.com/a/4105334/317623

jsarha replied:
It is just that Linux coding standard is the native language for so many of us here. The Linux coding standard does not forbid stack variable initializations, just mixing declarations (with or without initialization) and code, so it is possible satisfy both the security and the dinosaurs with more code blocks. Just saying, this is fine by me as it is.

marc-hb replied:
The Linux standard was stuck to C89 which forbid C99 mixing. It was recently upgraded to C11.

Leaving both security and dinosaurs aside for a minute, how would you rewrite this without mixing and without losing a few `const`? `{ }` blocks can often help but here these constants and variables are used in the entire function. So I'm afraid some constants would have to be downgraded to variables, no?

jsarha replied:
I do not see the problem, am I missing something?

`static void open_file(const char *watched_dir, const char *expected_file)
{
	const int iqueue = inotify_init();
	const int dwatch = inotify_add_watch(iqueue, watched_dir, IN_CREATE);
	struct stat expected_stat;
	void *ret_stream = NULL;
	char * const fpath = malloc(strlen(watched_dir) + 1 + strlen(expected_file) + 1);

...
	return ret_stream;
}


static void *wait_open(const char *watched_dir, const char *expected_file)
{
	if (access(watched_dir, R_OK)) {
		fprintf(stderr, "error: %s() cannot read %s\n", __func__, watched_dir);
		return NULL;
	}

	return open_files(watched_dir, expected_file);
}`

marc-hb replied:
Agreed, you can always add extra functions until mixing is avoided, artificially creating as many blocks as you need. However having a tiny "public" function that merely checks whether an argument exists and then delegates all the rest to another function that actually does all the work is an awkward solution for a C89 problem that does not exist anymore.

BTW many people don't even understand this old C89 rule, in the past I had review comments asking me not to do this:
```C
f(...)
{
   statement1;
    { int i;

    }
}

```

Don't get me wrong: giant functions are bad too but that's off topic.

ujfalusi commented on tools/logger/logger.c:
```diff
@@ -171,6 +174,75 @@ static int append_filter_config(struct convert_config *config, const char *input
+
+	const int iqueue = inotify_init();
+	const int dwatch = inotify_add_watch(iqueue, watched_dir, IN_CREATE);
+	FILE *fstream;
```
If you initialize it to NULL then you don't need to do it in all error paths?

ujfalusi commented on tools/logger/logger.c:
```diff
@@ -339,7 +411,15 @@ int main(int argc, char *argv[])
+			config.in_fd = fopen(config.in_file, "rb");
+		else
+			config.in_fd = wait_fopen(SYS_DEBUG,
+						  config.in_file + strlen(SYS_DEBUG) + 1);
```
I would be tempted to move this whole logic down into a custom `logger_fopen()` or something.
That way you would not need to re-construct the config.in_file into a new malloced buffer?

marc-hb replied:
Future possible improvement? Keep in mind we want to discard this logger entirely in the longer term.

ujfalusi reviewed (changes requested):
@marc-hb, this is a great idea, but I think the implementation can be made a bit more readable

kv2019i commented on tools/logger/logger.c:
```diff
@@ -171,6 +174,75 @@ static int append_filter_config(struct convert_config *config, const char *input
+		struct pollfd pfd[1] = {{ .fd = iqueue, .events = POLLIN, .revents = 0 }};
+		int pret = poll(pfd, sizeof(pfd) / sizeof(struct pollfd), timeout_min * 60 * 1000);
+
+		if (pret == 0) {
```
Maybe have a define for the "5" minutes. Or maybe just wait indefinitely? I mean if someone depends on error being returned if driver doesn't show up, 5 minutes is an awful lot of time to wait for (e.g. for test automation). So it could might as well be infinite wait.

marc-hb replied:
This code started with an infinite wait but then I realized a serious backward-compatibility issue: some test code could already be trying to start the logger before the driver is loaded (and ignoring the error as usual in shell scripts or C).

kv2019i reviewed (commented):
Thanks @marc-hb ! A few minor comments, but the main logic seems solid.

marc-hb commented on tools/logger/logger.c:
```diff
@@ -170,6 +174,75 @@ static int append_filter_config(struct convert_config *config, const char *input
+		 * used by some distros but broken in gcc:
+		 * https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425
+		 */
+		if (read(iqueue, evlist, sizeof(evlist)))
```
```suggestion
		(void) !read(iqueue, evlist, sizeof(evlist)))
```

lgirdwood reviewed (commented):
Lets get this in v2.0

lyakh commented on tools/logger/logger.c:
```diff
@@ -171,6 +174,75 @@ static int append_filter_config(struct convert_config *config, const char *input
+	fflush(stdout);
+
+	while (1) {
+		char evlist[1 * sizeof(struct inotify_event)];
```
mmh, how about `struct inotify_event event;`?

marc-hb replied:
See below.

lyakh commented on tools/logger/logger.c:
```diff
@@ -171,6 +174,75 @@ static int append_filter_config(struct convert_config *config, const char *input
+	while (1) {
+		char evlist[1 * sizeof(struct inotify_event)];
+		const int timeout_min = 5;
+		struct pollfd pfd[1] = {{ .fd = iqueue, .events = POLLIN, .revents = 0 }};
```
`struct pollfd pfd` would do too

marc-hb replied:
I prefer to stick to the exact API signatures and not confuse arrays with their first element. It's also more future proof.

gkbldcig commented:
Can one of the admins verify this patch?

marc-hb reviewed (commented):
I've been testing this and it has been working fine for me but I noticed logical error. Right now it's using inotify to watch `/sys/kernel/debug` until it `/sys/kernel/debug/sof/[e]trace` exists. However inotify is not recursive. So, it gets notified when `/sys/kernel/debug/sof` gets created, but if `/sys/kernel/debug/sof/[e]trace` gets created long enough after `/sys/kernel/debug/sof`  is then it could hang forever. Any idea why it seemed to be working fine so far?

lgirdwood commented:
> I've been testing this and it has been working fine for me but I noticed a logical error. Right now it's using inotify to watch `/sys/kernel/debug` until `/sys/kernel/debug/sof/[e]trace` exists. However inotify is not recursive. So, it gets notified when `/sys/kernel/debug/sof` gets created, but if `/sys/kernel/debug/sof/[e]trace` gets created long enough after `/sys/kernel/debug/sof` then the logger could hang forever. Any /sys idea why it seems to be working fine so far anyway? A race condition maybe?

Does it make sense to simplify and monitor only 1 file per process i.e. logger can be invoked with either etrace or trace arguments. This would simpligy, any output can always be stitched together later on with a script.

marc-hb commented:
> logger can be invoked with either etrace or trace arguments

This is already the case, I think you didn't get my concern.

My concern is this PR works thanks to a race.

- `/sys/kernel/debug/sof/` gets created, sof-logger gets notified and looks for `/sys/kernel/debug/sof/trace`, does not find it YET.
- A LONG enough time later, `/sys/kernel/debug/sof/trace` gets created but that's one level too far down so NO `/sys/kernel/debug/` notification. `sof-logger` is stuck forever.

I will experiment a bit more.

lgirdwood reviewed (commented):
@marc-hb can you give us an update about what has changed since last push (as some time has passes and many changes have been made to align the logging with Zephyr). What is the plan here ?

marc-hb commented:
> My concern is this PR works thanks to a race.

I fixed that some time ago but never shared, now I did. I also addressed some minor review comments.

I've been testing this PR in my workspace for months, it's very solid.
